Reset parallel test tables by deleting

djmb opened the pull request:
### Motivation / Background

Switch from truncating to deleting when resetting parallel test databases. This PR is an alternative to https://github.com/rails/rails/pull/56026 without the configuration option.

When doing parallel testing with many databases, truncating all the tables is very slow on MySQL. Deleting from the tables is much faster even if they contain a reasonable amount of data.

### Detail

Running on a schema with about 200 empty tables and 32 parallel databases, deleting from all tables takes about 0.3 seconds, while truncating takes 18 seconds. Even with 1000 rows to delete in each table deleting is about 2.7x faster than truncating.

On PostgreSQL, truncating the tables is slower than deleting if they have small amounts of data which is typical for test databases. Both methods are very quick anyway.

There's no difference between the two methods for SQLite, as the adapter uses DELETE for truncation anyway.

Functionally DELETE and TRUNCATE are identical on PostgreSQL, but on MySQL TRUNCATE will reset AUTO_INCREMENT counters while DELETE will not.

Tests should not be relying on them being reset and since this is only for parallel tests where test orders are going to be randomised anyway so the IDs are unlikely to be consistent between test runs anyway.

By switching to DELETE all three databases will behave identically and not reset the ID sequence counters.

If there's much more data in the tables then TRUNCATE may become faster, but in that case the tests are likely to be slow as they'll need to generate the data.

This PR also removed the SKIP_TEST_DATABASE_TRUNCATE environment variable - we no longer truncate so the name doesn't make sense and deleting should be fast enough to not need to skip.


### Additional information

Timings for resetting tables with TRUNCATE vs DELETE, based on 32 databases with 200 tables each, with 0, 100 and 1000 rows per table:
```
  ┌──────────────┬─────────┬──────────┬───────┬───────┬───────┬─────────┐
  │   Database   │  Rows/  │  Method  │  Min  │  Avg  │  Max  │ Winner  │
  │              │  Table  │          │ (sec) │ (sec) │ (sec) │         │
  ├──────────────┼─────────┼──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ TRUNCATE │ 0.269 │ 0.535 │ 0.578 │         │
  │ PostgreSQL   │    0    │──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ DELETE   │ 0.038 │ 0.051 │ 0.067 │  10.6x  │
  ├──────────────┼─────────┼──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ TRUNCATE │ 0.174 │ 0.323 │ 0.408 │         │
  │ PostgreSQL   │   100   │──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ DELETE   │ 0.104 │ 0.132 │ 0.150 │  2.45x  │
  ├──────────────┼─────────┼──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ TRUNCATE │ 0.511 │ 0.578 │ 0.604 │  2.48x  │
  │ PostgreSQL   │  1000   │──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ DELETE   │ 1.372 │ 1.434 │ 1.568 │         │
  ├──────────────┼─────────┼──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ TRUNCATE │17.485 │18.413 │18.978 │         │
  │ MySQL        │    0    │──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ DELETE   │ 0.311 │ 0.341 │ 0.356 │  53.6x  │
  ├──────────────┼─────────┼──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ TRUNCATE │16.892 │18.281 │18.845 │         │
  │ MySQL        │   100   │──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ DELETE   │ 0.534 │ 1.406 │ 2.003 │  13.0x  │
  ├──────────────┼─────────┼──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ TRUNCATE │16.723 │17.939 │18.544 │         │
  │ MySQL        │  1000   │──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ DELETE   │ 3.789 │ 6.635 │ 8.400 │  2.70x  │
  └──────────────┴─────────┴──────────┴───────┴───────┴───────┴─────────┘
```

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

djmb committed:
Reset parallel test tables by deleting

When doing parallel testing with many databases, truncating all the
tables is very slow on MySQL. Deleting from the tables is much faster
even if they contain a reasonable amount of data.

Running on a schema with about 200 empty tables and 32 parallel
databases, deleting from all tables takes about 0.3 seconds, while
truncating takes 18 seconds. Even with 1000 rows to delete in each
table deleting is about 2.7x faster than truncating.

On PostgreSQL, truncating the tables is slower than deleting if they
have small amounts of data which is typical for test databases. Both
methods are very quick anyway.

There's no difference between the two methods for SQLite, as the adapter
uses DELETE for truncation anyway.

Functionally DELETE and TRUNCATE are identical on PostgreSQL, but on
MySQL TRUNCATE will reset AUTO_INCREMENT counters while DELETE will not.

Tests should not be relying on them being reset and since this is only
for parallel tests where test orders are going to be randomised anyway
so the IDs are unlikely to be consistent between test runs anyway.

By switching to DELETE all three databases will behave identically
and not reset the ID sequence counters.

Timings for resetting tables with TRUNCATE vs DELETE, based on 32
databases with 200 tables each, with 0, 100 and 1000 rows per table:
```
  ┌──────────────┬─────────┬──────────┬───────┬───────┬───────┬─────────┐
  │   Database   │  Rows/  │  Method  │  Min  │  Avg  │  Max  │ Winner  │
  │              │  Table  │          │ (sec) │ (sec) │ (sec) │         │
  ├──────────────┼─────────┼──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ TRUNCATE │ 0.269 │ 0.535 │ 0.578 │         │
  │ PostgreSQL   │    0    │──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ DELETE   │ 0.038 │ 0.051 │ 0.067 │  10.6x  │
  ├──────────────┼─────────┼──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ TRUNCATE │ 0.174 │ 0.323 │ 0.408 │         │
  │ PostgreSQL   │   100   │──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ DELETE   │ 0.104 │ 0.132 │ 0.150 │  2.45x  │
  ├──────────────┼─────────┼──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ TRUNCATE │ 0.511 │ 0.578 │ 0.604 │  2.48x  │
  │ PostgreSQL   │  1000   │──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ DELETE   │ 1.372 │ 1.434 │ 1.568 │         │
  ├──────────────┼─────────┼──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ TRUNCATE │17.485 │18.413 │18.978 │         │
  │ MySQL        │    0    │──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ DELETE   │ 0.311 │ 0.341 │ 0.356 │  53.6x  │
  ├──────────────┼─────────┼──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ TRUNCATE │16.892 │18.281 │18.845 │         │
  │ MySQL        │   100   │──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ DELETE   │ 0.534 │ 1.406 │ 2.003 │  13.0x  │
  ├──────────────┼─────────┼──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ TRUNCATE │16.723 │17.939 │18.544 │         │
  │ MySQL        │  1000   │──────────┼───────┼───────┼───────┼─────────┤
  │              │         │ DELETE   │ 3.789 │ 6.635 │ 8.400 │  2.70x  │
  └──────────────┴─────────┴──────────┴───────┴───────┴───────┴─────────┘
```

If there's much more data in the tables then TRUNCATE may become
faster, but in that case the tests are likely to be slow as they'll need
to generate the data.

This PR also removed the SKIP_TEST_DATABASE_TRUNCATE environment
variable - we no longer truncate so the name doesn't make sense and
deleting should be fast enough to not need to skip.

byroot commented:
Also, a bit tangential, but when we look at this at Shopify, at some point rather than to empty table on test start, we did attempt to checksum the tables to ensure the inserted fixtures were still valid. Problem is it doesn't play very well with dybamic fixtures such as `created_at: <%= 10.minutes.ago %>`.

djmb commented:
@byroot - I tested against MySQL 8.0.43, and Postgres 18.0.

Adding `empty_all_tables` to the adapter seems a bit dangerous, but I suppose there's already `drop_database` there if you are looking for footguns.

If we go with that, we could stick with truncate for Postgres I guess? It doesn't matter a huge amount, either way is pretty fast. 

It would also mean we don't have to use send to call `execute_batch`, though maybe it should be a public method anyway - it seems like something that could be generally useful?

djmb committed:
Add AbstractAdapter#empty_all_tables

Implement MySQLAdapter#empty_all_tables to use DELETE instead
of TRUNCATE. Keep truncating for other adapters (though SQLite
implements truncation with DELETE).

djmb commented:
@byroot - thanks for the review. I've added `empty_all_tables` to the `AbstractAdapter` now and defaulted to truncating. I've only changed the MySQL adapter now to use delete - there's not much difference between delete and truncate on Postgresql.

This all means that also we won't automatically change how it works for third part adapters which is probably for the best.
